// meteor-desktop/dist/installerBuilder.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _electronBuilder = require('electron-builder');

var _shelljs = require('shelljs');

var _shelljs2 = _interopRequireDefault(_shelljs);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _log = require('./log');

var _log2 = _interopRequireDefault(_log);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Wrapper for electron-builder.
 */
var InstallerBuilder = function () {

    /**
     * @param {MeteorDesktop} $ - context
     *
     * @constructor
     */
    function InstallerBuilder($) {
        (0, _classCallCheck3.default)(this, InstallerBuilder);

        this.log = new _log2.default('installerBuilder');
        this.$ = $;
    }

    (0, _createClass3.default)(InstallerBuilder, [{
        key: 'build',
        value: function () {
            var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee() {
                var settings, arch, targets, target;
                return _regenerator2.default.wrap(function _callee$(_context) {
                    while (1) {
                        switch (_context.prev = _context.next) {
                            case 0:
                                // Move node_modules away. We do not want to delete it, just temporarily remove it from
                                // our way.
                                _shelljs2.default.mv(this.$.env.paths.electronApp.nodeModules, this.$.env.paths.electronApp.tmpNodeModules);

                                settings = this.$.desktop.getSettings();

                                if (!('builderOptions' in settings)) {
                                    this.log.error('no builderOptions in settings.json, aborting');
                                    process.exit(1);
                                }

                                // We are handling asar'ing and rebuilding in the normal run/build flow so we do not
                                // want electron-rebuild to do that.
                                settings.builderOptions.asar = false;
                                settings.builderOptions.npmRebuild = true;

                                arch = this.$.env.options.ia32 ? 'ia32' : 'x64';


                                arch = this.$.env.options.allArchs ? 'all' : arch;

                                targets = [];


                                if (this.$.env.options.win) {
                                    targets.push(_electronBuilder.Platform.WINDOWS);
                                }
                                if (this.$.env.options.linux) {
                                    targets.push(_electronBuilder.Platform.LINUX);
                                }
                                if (this.$.env.options.mac) {
                                    targets.push(_electronBuilder.Platform.MAC);
                                }

                                if (targets.length === 0) {
                                    if (this.$.env.os.isWindows) {
                                        targets.push(_electronBuilder.Platform.WINDOWS);
                                    } else if (this.$.env.os.isLinux) {
                                        targets.push(_electronBuilder.Platform.LINUX);
                                    } else {
                                        targets.push(_electronBuilder.Platform.MAC);
                                    }
                                }

                                target = (0, _electronBuilder.createTargets)(targets, null, arch);
                                _context.prev = 13;
                                _context.next = 16;
                                return (0, _electronBuilder.build)({
                                    targets: target,
                                    devMetadata: {
                                        directories: {
                                            app: this.$.env.paths.electronApp.root,
                                            output: _path2.default.join(this.$.env.options.output, this.$.env.paths.installerDir)
                                        },
                                        build: settings.builderOptions
                                    }
                                });

                            case 16:
                                _context.next = 21;
                                break;

                            case 18:
                                _context.prev = 18;
                                _context.t0 = _context['catch'](13);

                                this.log.error('error while building installer: ', _context.t0);

                            case 21:
                                _context.prev = 21;

                                // Move node_modules back.
                                _shelljs2.default.mv(this.$.env.paths.electronApp.tmpNodeModules, this.$.env.paths.electronApp.nodeModules);
                                return _context.finish(21);

                            case 24:
                            case 'end':
                                return _context.stop();
                        }
                    }
                }, _callee, this, [[13, 18, 21, 24]]);
            }));

            function build() {
                return _ref.apply(this, arguments);
            }

            return build;
        }()
    }]);
    return InstallerBuilder;
}();

exports.default = InstallerBuilder;
